File: typescript/src/problems201_300/invert-binary-tree.ts
import TreeNode from '../concepts/tree-node';

class Problem {
  invertTree(root: TreeNode | undefined): TreeNode | undefined {
    throw new Error(`Not implemented (${root}).`);
  }
}

class Default extends Problem {
  invertTree(root: TreeNode | undefined): TreeNode | undefined {
    if (!root) {
      return undefined;
    }
    let temp = root.left;
    root.left = this.invertTree(root.right);
    root.right = this.invertTree(temp);
    return root;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems201_300/invert_binary_tree.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.tree_node import TreeNode


class Problem(ABC):
    @abstractmethod
    def invert_tree(self, root: TreeNode | None) -> TreeNode | None:
        pass


class Default(Problem):
    @override
    def invert_tree(self, root: TreeNode | None) -> TreeNode | None:
        if not root:
            return None
        temp: TreeNode | None = root.left
        root.left = self.invert_tree(root.right)
        root.right = self.invert_tree(temp)
        return root


SOLUTIONS: list[Problem] = [Default()]
